builtin/providers/ultradns/resource_ultradns_record.go

```diff
@@ -0,0 +1,166 @@
+package ultradns
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/Ensighten/udnssdk"
+)
+
+func resourceUltraDNSRecord() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceUltraDNSRecordCreate,
+		Read:   resourceUltraDNSRecordRead,
+		Update: resourceUltraDNSRecordUpdate,
+		Delete: resourceUltraDNSRecordDelete,
+
+		Schema: map[string]*schema.Schema{
+			"zone": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"hostname": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"type": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"rdata": &schema.Schema{
+				Type:     schema.TypeList,
+				Required: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+
+			"ttl": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "3600",
+			},
+		},
+	}
+}
+
+func resourceUltraDNSRecordCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*udnssdk.Client)
+	newRecord := &udnssdk.RRSet{
+		OwnerName: d.Get("name").(string),
+		RRType:    d.Get("type").(string),
+	}
+	rdata := d.Get("rdata").([]interface{})
+	rdatas := make([]string, len(rdata))
+	for i, j := range rdata {
+		rdatas[i] = j.(string)
+	}
+	newRecord.RData = rdatas
+	if ttl, ok := d.GetOk("ttl"); ok {
+		newRecord.TTL, _ = strconv.Atoi(ttl.(string))
+	}
+
+	log.Printf("[DEBUG] UltraDNS RRSet create configuration: %#v", newRecord)
+
+	_, err := client.RRSets.CreateRRSet(d.Get("zone").(string), *newRecord)
+	recId := fmt.Sprintf("%s.%s", d.Get("name").(string), d.Get("zone").(string))
+	if err != nil {
+		return fmt.Errorf("Failed to create UltraDNS RRSet: %s", err)
+	}
+
+	d.SetId(recId)
+	log.Printf("[INFO] record ID: %s", d.Id())
+
+	return resourceUltraDNSRecordRead(d, meta)
+}
+
+func resourceUltraDNSRecordRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*udnssdk.Client)
+
+	rrsets, _, err := client.RRSets.GetRRSets(d.Get("zone").(string), d.Get("name").(string), d.Get("type").(string))
+	if err != nil {
+		return fmt.Errorf("Couldn't find UltraDNS RRSet: %s", err)
```

Comment:
We return general errors, but Not Found errors should be treated specially - doing `d.SetId("")` and `return nil` so Terraform knows that the record is gone and should be recreated to apply the config. This helps Terraform handle external deletion properly.

Here's an example of what I'm talking about from the AWS provider:

https://github.com/hashicorp/terraform/blob/4f035d444f0467d14cccdeac055ed1248955be4b/builtin/providers/aws/resource_aws_instance.go#L432-L442